x-pack/plugins/apm/server/routes/sampled_telemetry/route.ts

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { createOrUpdateIndex } from '@kbn/observability-plugin/server';
+import { createApmServerRoute } from '../apm_routes/create_apm_server_route';
+import { SavedServiceGroup } from '../../../common/service_groups';
+import { getDataTelemetry } from '../../lib/apm_telemetry';
+
+export const sampledTelemetryRoute = createApmServerRoute({
+  endpoint: 'GET /internal/apm/upload-sample-telemetry',
```

Comment:
You can add a simple api test for this endpoint. This way we'll have test coverage of our telemetry! :)

Reply:
What's the purpose of this endpoint? Is this a dummy thing or do we plan on shipping this to production?

Reply:
Plan on shipping it to production. Potentially lock down somehow if we need to.

Reply:
We're querying w/ the internal user which can read everything from the APM indices. The user that is accessing this endpoint might have more limited privileges. The index permissions for the write target could be different from the index permissions for the read targets. I think there's a risk of privilege escalation here - this data is usually only available to super users and the internal user, as the snapshot is stored in the `.kibana` index. Additionally, I'm not sure if creating an index for this purpose in a production environment is a good idea (e.g., this index will be included in searches to APM data indices, which might break searches or return incorrect data).

Reply:
Good points with privilege escalation. If we can avoid using the internal user and instead use the current users privileges we should do that. If not we should lock the endpoint down.

Reply:
> Additionally, I'm not sure if creating an index for this purpose in a production environment is a good idea (e.g., this index will be included in searches to APM data indices, which might break searches or return incorrect data).

Again, we can lock this more down if we deem it necessary - but I really want to avoid this. The purpose if this endpoint is to make it much more transparent what telemetry we are collecting, and lowering the barriers to see it and build visualisations for it.

We can change the prefix from "apm" but it shouldn't be needed. It was needed years ago but we introduced `processor.event` filters for this exact reason. Also, if we don't document this API publicly how would customers hit this?

Reply:
> Again, we can lock this more down if we deem it necessary - but I really want to avoid this. The purpose if this endpoint is to make it much more transparent what telemetry we are collecting, and lowering the barriers to see it and build visualisations for it.

This data is stored in the .kibana index, and users can already build visualisations against it (albeit w/ runtime fields in some cases). Is adding an endpoint w/ the potential of privilege escalation and copying over data to another index that has to be managed worth the ease-of-use of an explicit endpoint?


fwiw, I'm not actually sure if this is a privilege escalation. Telemetry data is already sent from the browser. I'm just not sure whether that happens w/ any user and whether or not that data is encrypted. Additionally, I'm not sure if creating an index for this purpose in a production environment is a good idea (e.g., this index will be included in searches to APM data indices, which might break searches or return incorrect data).

> We can change the prefix from "apm" but it shouldn't be needed. It was needed years ago but we introduced processor.event filters for this exact reason. Also, if we don't document this API publicly how would customers hit this accidentally?

There might be (or are) cases where we query *apm* indices/data streams but don't use `apm.events`, e.g. when we use an ES API for which there is no equivalent in the APMEventClient.

Reply:
> Is adding an endpoint w/ the potential of privilege escalation and copying over data to another index that has to be managed worth the ease-of-use of an explicit endpoint?

If we change calls to be made by the current user, there'd be no privilege escalation concerns, right? And yes, I think adding runtime fields is a bit too much. Preferably it is possible to work with telemetry data out of the box (which is why i think we should also create a data view automatically).

>  I'm not sure if creating an index for this purpose in a production environment is a good idea 

Again, how will this accidentally be called in a production environment? Do we expect users to randomly be hitting endpoints? Even if this _does_ happen by a legitimate accident it shouldn't have any consequence on search queries. And even in the edge case that it does, the remedy is very simple: delete the index. If we were deleting data or doing anything irreversible I'd understand the concern but the possible negative impact in this case seems rather low.

Reply:
> If we change calls to be made by the current user, there'd be no privilege escalation concerns, right? And yes, I think adding runtime fields is a bit too much. Preferably it is possible to work with telemetry data out of the box (which is why i think we should also create a data view automatically).

There is, because we're storing data in a separate index where certain index-level or document-level privileges that are applicable to APM data indices do not apply. Re: runtime fields, I think dynamic index mappings will end up being not usable either, esp. w/ the planned changes (e.g. service-level telemetry). Also, the data from other plugins is not included here and cannot be aggregated together.

> Again, how will this accidentally be called in a production environment? Do we expect users to randomly be hitting endpoints? Even if this does happen by a legitimate accident it shouldn't have any consequence on search queries. And even in the edge case that it does, the remedy is very simple: delete the index. If we were deleting data or doing anything irreversible I'd understand the concern but the possible negative impact in this case seems rather low.

Agree that it's unlikely, I think I'm mostly sceptical about the value vs the downsides.

Reply:
> There is, because we're storing data in a separate index where certain index-level or document-level privileges that are applicable to APM data indices do not apply.

Are you concerned that a superuser with legitimate access accidentally accesses the endpoint thereby making telemetry available to users who shouldn't be able to see it?
Or are you concerned that malicious users can use the endpoint to access data they shouldn't have access to in the first place?

> Agree that it's unlikely, I think I'm mostly sceptical about the value vs the downsides

The value is yet to be seen. I'm optimistic but if it's not useful we'll just remove it again (it needs to be secure ofc)

Reply:
> Are you concerned that a superuser accidentally accesses the endpoint thereby making telemetry data available to users who shouldn't be able to see it?
> Or are you concerned that malicious users can use the endpoint to access data they shouldn't have access to in the first place?

The former, the latter is no longer applicable if all data is queried with the current user rather than the internal user. Who would access these endpoints in production?